Custom test framework

paolochiodi opened the pull request:
I opened this PR against my other PR so that changes to the code are clearer and it's hopefully easier to check if the new framework brings more clarity.

This PR currently contains the first step of such a framework: a function to build all initial records in one place. It is currently called `prepareTest` and returns an object intended to be later used to access created records (and in further iteration also dry-up the test definitions).

Two thing to note:
- Created records can be accessed by the key used in the records definition, but will have generic auto-generated ids
- Internally it still uses `lab.before` and `lab.after`, which I think is good for keeping everything in sync

At this time I'm seeking feedbacks on the type of approach and the "api definition" of the `prepareTest` function, basically this piece of code:

```js
      const test = prepareTest(lab, {
        teams: {
          calledTeam: { name: 'called team', organizationId, users: ['called'] }
        },
        users: {
          caller: { name: 'caller', organizationId, policies: ['testedPolicy'] },
          called: { name: 'called', organizationId }
        },
        policies: {
          testedPolicy: Policy()
        }
       })
```

Please go light on the implementation has it is still temporary to demonstrate my intentions (it will probably be moved to utils)

Next steps will be:
- use `prepareTest` in all experiments in this file, expand as needed
- create framework to define auth tests, something along the lines of
```js
      test('should authorize caller with policy for specific users')
        .withPolicy([{
          Effect: 'Allow',
          Action: ['authorization:users:read'],
          Resource: [`/authorization/user/WONKA/*/${calledId}`]
        }])
        .endpoint({
          method: 'GET',
          url: `/authorization/users/${calledId}`,
          headers: { authorization: callerId }
        })
        .shouldRespond(200)
```

nherment commented:
@paolochiodi looks good ! I would be even more aggressive with the verbosity (ie get rid of all the js/json-like objects) but it's nitpicking.

p16 commented:
@paolochiodi I think the direction is the right one. I'm not really sure about the amount of code we need to write to load the fixtures we want. Do you think we can use something like [`node-sql-fixtures`](https://github.com/city41/node-sql-fixtures) ?

The end goal I'd like to achieve is having the fixtures loaded without necessarily depend on our own ops functions. This may also save us the trouble of changing a lot of tests if we change the interface of a function.

I'm not sure if this is the place to discuss this (we also have a ticket assigned to @nherment https://github.com/nearform/labs-authorization/issues/254 )

paolochiodi commented:
@p16 Yeah, we may end up using something like that.  
As a first step I just wanted to extract the code we are using now. 
If we expand the usage of "prepareTest" we can later rewrite its implementation to whatever fits better.

About using our own *Ops functions: for the moment it has the benefits of managing all the side effects (i.e. will delete team_users when a team is deleted).
I agree that in the long run this may not be the best solution, because the functions can get more complex and maybe have side-effects we don't want to replicate (i.e. creating an audit log, sending email etc)

mihaidma commented:
Is it worth considering clean up for the whole db?

My use case scenario: I do a debug, stop debug at breakpoint, fix, npm run init db, npm run load test data, npm run again debug or tests. Generating new ids at each run avoids this type of problems, no not a big issue.

I like the test structure, looks very good.

paolochiodi commented:
@nherment @p16 @mihaidma I've updated the PR extending the usage of prepareTest to other tests. (I had to expand it a bit to accomodate all use cases).

I've also introduced some changes:
- prepareTest is now Factory
- the factory now returns the object with the records directly
- other minor name changes

Let me know if you have any additional feedback.
Before the end of the day I'll close this PR and merge it into the main PR, where I'll address any further change.
